File: src/day2/RotateArray.java
package day2;

public class RotateArray {
    public static void main(String[] args) {
        int[] nums = {1, 2, 3};
        rotate(nums, 4);

        for (int num : nums) {
            System.out.println("num = " + num);
        }
    }

    public static void rotate(int[] nums, int k) {
        if (nums.length < k) {
            k = k % nums.length;
        }
        reverse(nums, 0, nums.length - 1);
        reverse(nums, 0, k - 1);
        reverse(nums, k, nums.length - 1);
    }

    public static void reverse(int[] nums, int left, int right) {
        while (left <= right) {
            int temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;
            left++;
            right--;
        }
    }
}

File: src/kotlinsource/day2/RotateArray.kt
package kotlinsource.day2

class RotateArray {
    fun rotate(nums: IntArray, k: Int): Unit {
        var target: Int = k
        if (nums.size < target) {
            target %= nums.size
        }

        reverse(nums, 0, nums.size - 1)
        reverse(nums, 0, target - 1)
        reverse(nums, target, nums.size - 1)
    }

    private fun reverse(nums: IntArray, i: Int, j: Int) {
        var left = i
        var right = j
        while (left <= right) {
            val temp: Int = nums[left]
            nums[left] = nums[right]
            nums[right] = temp
            left--
            right++
        }
    }
}

fun main() {

}